plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSourceRecordProcessor.java

```diff
@@ -0,0 +1,269 @@
+			if(this.connection != null) {
+				// Prepare the request.
+				if (!connection.getMetadata().canRead()) {
+					throw new ProcessException("Reading not supported by connection");
+				}
+				String inputFileUUID = fileToProcess == null ? null : fileToProcess.getAttribute(CoreAttributes.UUID.key());
+				Map<String, String> inputFileAttrMap = fileToProcess == null ? null : fileToProcess.getAttributes();
+				FlowFile resultSetFF;
+				if (fileToProcess == null) {
+					resultSetFF = session.create();
+				} else {
+					resultSetFF = session.create(fileToProcess);
+				}
+				if (inputFileAttrMap != null) {
+					resultSetFF = session.putAllAttributes(resultSetFF, inputFileAttrMap);
+				}
+				try {
+					PlcReadRequest.Builder builder = connection.readRequestBuilder();
+					getFields().forEach(field -> {
+						String address = getAddress(field);
+						if (address != null) {
+							builder.addItem(field, address);
+						}
+					});
+					PlcReadRequest readRequest = builder.build();
+					PlcReadResponse readResponse = readRequest.execute().get(10, TimeUnit.SECONDS);
+					resultSetFF = session.write(resultSetFF, out -> {
+						try {
+							nrOfRows.set(plc4xWriter.writePlcReadResponse(readResponse, this.getPlcAddress(), out, logger, null, PROTOCOL));
+						} catch (Exception e) {
+							throw (e instanceof ProcessException) ? (ProcessException) e : new ProcessException(e);
+						}
+					});
+					long executionTimeElapsed = executeTime.getElapsed(TimeUnit.MILLISECONDS);
+					final Map<String, String> attributesToAdd = new HashMap<>();
+					attributesToAdd.put(RESULT_ROW_COUNT, String.valueOf(nrOfRows.get()));
+					attributesToAdd.put(RESULT_QUERY_EXECUTION_TIME, String.valueOf(executionTimeElapsed));
+					if (inputFileUUID != null) {
+						attributesToAdd.put(INPUT_FLOWFILE_UUID, inputFileUUID);
+					}
+					attributesToAdd.putAll(plc4xWriter.getAttributesToAdd());
+					resultSetFF = session.putAllAttributes(resultSetFF, attributesToAdd);
+					plc4xWriter.updateCounters(session);
+					logger.info("{} contains {} records; transferring to 'success'",
+							new Object[] { resultSetFF, nrOfRows.get() });
+					// Report a FETCH event if there was an incoming flow file, or a RECEIVE event otherwise
+					if (context.hasIncomingConnection()) {
+						session.getProvenanceReporter().fetch(resultSetFF, "Retrieved " + nrOfRows.get() + " rows", executionTimeElapsed);
+					} else {
+						session.getProvenanceReporter().receive(resultSetFF, "Retrieved " + nrOfRows.get() + " rows", executionTimeElapsed);
+					}
+					resultSetFlowFiles.add(resultSetFF);
+					if (resultSetFlowFiles.size() >= 0) {
+						session.transfer(resultSetFlowFiles, super.REL_SUCCESS);
+						// Need to remove the original input file if it exists
+						if (fileToProcess != null) {
+							session.remove(fileToProcess);
+							fileToProcess = null;
+						}
+						session.commit();
+						resultSetFlowFiles.clear();
+					}
+				} catch (Exception e) {
```

Comment:
English please

Reply:
This logic around the try catches seems involved, perhaps a comment explaining what you are doing and why would help maintainability